In [74]:
import pandas as pd
import numpy as np
pd.set_option('display.max_columns', None)
pd.set_option('display.float_format', '{:,.0f}'.format)

import warnings
warnings.filterwarnings("ignore")

In [75]:
df = pd.read_csv('../data/cleaned_data.csv')

df['date'] = pd.to_datetime(df['date'])
df.head()

date  store_nbr      family  sales  onpromotion holiday_type  \
0 2013-01-01          1  AUTOMOTIVE      0            0      Holiday   
1 2013-01-01          1   BABY CARE      0            0      Holiday   
2 2013-01-01          1      BEAUTY      0            0      Holiday   
3 2013-01-01          1   BEVERAGES      0            0      Holiday   
4 2013-01-01          1       BOOKS      0            0      Holiday   

     locale  transferred  dcoilwtico   city      state store_type  cluster  \
0  National        False          93  Quito  Pichincha          D       13   
1  National        False          93  Quito  Pichincha          D       13   
2  National        False          93  Quito  Pichincha          D       13   
3  National        False          93  Quito  Pichincha          D       13   
4  National        False          93  Quito  Pichincha          D       13   

   transactions  year  month  week  quarter day_of_week  is_crisis  \
0             0  2013      1     1        1     Tuesday          0   
1             0  2013      1     1        1     Tuesday          0   
2             0  2013      1     1        1     Tuesday          0   
3             0  2013      1     1        1     Tuesday          0   
4             0  2013      1     1        1     Tuesday          0   

   sales_lag_7  rolling_mean_7  is_weekend  is_holiday  promo_last_7_days  
0            0               0           0           1                  0  
1            0               0           0           1                  0  
2            0               0           0           1                  0  
3            0               0           0           1                  0  
4            0               0           0           1                  0

## 🔍 Question 1: How Have Total Sales Evolved Over Time?

To understand the overall business trend, we calculated the total sales per day from the dataset.

In [76]:
sales_over_time = df.groupby('date')['sales'].sum().reset_index()
sales_over_time

date   sales
0    2013-01-01   2,512
1    2013-01-02 496,092
2    2013-01-03 361,461
3    2013-01-04 354,460
4    2013-01-05 477,350
...         ...     ...
1679 2017-08-11 826,374
1680 2017-08-12 792,631
1681 2017-08-13 865,640
1682 2017-08-14 760,922
1683 2017-08-15 762,662

[1684 rows x 2 columns]

### Key Findings:
- Daily sales range from as low as ~2.5K to over 860K in some peak days.
- There is a clear upward trend in daily revenue, with seasonal fluctuations likely present (to be analyzed in later steps).

> Corresponding plots will be in cell `4` in `visualization_demo.ipynb`

## 🔍 2. Which products or categories contribute the most to total revenue?

Based on the total sales data, the following products or categories contribute the most to the total revenue:

In [77]:
top_products = df.groupby('family')['sales'].sum().sort_values(ascending=False).head(20)
top_products

family
GROCERY I             350,827,298
BEVERAGES             221,663,540
PRODUCE               125,447,968
CLEANING               99,421,019
DAIRY                  65,823,605
BREAD/BAKERY           42,959,924
POULTRY                32,494,451
MEATS                  31,650,996
PERSONAL CARE          25,100,482
DELI                   24,585,627
HOME CARE              16,409,522
EGGS                   15,881,196
FROZEN FOODS           14,646,940
PREPARED FOODS          8,966,728
LIQUOR,WINE,BEER        7,937,172
SEAFOOD                 2,051,636
GROCERY II              2,004,966
HOME AND KITCHEN I      1,905,076
HOME AND KITCHEN II     1,556,511
CELEBRATION               779,502
Name: sales, dtype: float64

1. **GROCERY I**: $350,827,298
2. **BEVERAGES**: $221,663,540
3. **PRODUCE**: $125,447,968
4. **CLEANING**: $99,421,019
5. **DAIRY**: $65,823,605

These categories make up the bulk of the revenue, with **GROCERY I** leading by a significant margin. The top five categories contribute substantially to the overall sales, while the remaining categories (such as **CELEBRATION** and **HOME AND KITCHEN II**) have relatively smaller contributions.

In the analysis, we can observe that categories related to essential products (like groceries, beverages, and produce) lead in sales, which might reflect consistent consumer demand. Further analysis could explore seasonality and trends within these top categories.

> Corresponding plots will be in cell `6` in `visualization_demo.ipynb`

## 🔍 3. Which stores, cities, or states are the top performers in terms of revenue?

In [78]:
top_stores = df.groupby('store_nbr')['sales'].sum().sort_values(ascending=False)
top_cities = df.groupby('city')['sales'].sum().sort_values(ascending=False)
top_regions = df.groupby('state')['sales'].sum().sort_values(ascending=False)

print("Top Stores by Revenue:")
print(top_stores.head())  

print("\n \nTop Cities by Revenue:")
print(top_cities.head()) 

print("\n \nTop States by Revenue:")
print(top_regions.head()) 

Top Stores by Revenue:
store_nbr
44   63,356,137
45   55,689,022
47   52,024,476
3    51,533,528
49   44,346,823
Name: sales, dtype: float64

 
Top Cities by Revenue:
city
Quito           568,679,349
Guayaquil       125,572,186
Cuenca           50,194,046
Ambato           41,159,773
Santo Domingo    36,617,572
Name: sales, dtype: float64

 
Top States by Revenue:
state
Pichincha                        597,585,883
Guayas                           168,649,985
Azuay                             50,194,046
Tungurahua                        41,159,773
Santo Domingo de los Tsachilas    36,617,572
Name: sales, dtype: float64


Based on the total sales data, the following stores, cities, and regions are the top performers:

### **Top Stores by Revenue:**
1. **Store 44**: $63,356,137
2. **Store 45**: $55,689,022
3. **Store 47**: $52,024,476
4. **Store 3**: $51,533,528
5. **Store 49**: $44,346,823

### **Top Cities by Revenue:**
1. **Quito**: $568,679,349
2. **Guayaquil**: $125,572,186
3. **Cuenca**: $50,194,046
4. **Ambato**: $41,159,773
5. **Santo Domingo**: $36,617,572

### **Top States by Revenue:**
1. **Pichincha**: $597,585,883
2. **Guayas**: $168,649,985
3. **Azuay**: $50,194,046
4. **Tungurahua**: $41,159,773
5. **Santo Domingo de los Tsachilas**: $36,617,572

These top performers highlight the most significant contributors to revenue, with **Quito** leading at the city level and **Pichincha** being the highest-performing state. In terms of stores, Store 44 generates the highest revenue.

This analysis can help identify key areas for growth and focus, particularly in high-revenue cities and states.

## 🔍 4. What is the average order size across stores, regions, and categories?

In [79]:
df['transactions'].dtype

dtype('float64')

In [80]:
df_filtered = df[df['transactions'] > 0]

avg_order_size_store = df_filtered.groupby('store_nbr').apply(lambda x: x['sales'].sum() / x['transactions'].sum()).sort_values(ascending=False)
avg_order_size_region = df_filtered.groupby('state').apply(lambda x: x['sales'].sum() / x['transactions'].sum()).sort_values(ascending=False)
avg_order_size_category = df_filtered.groupby('family').apply(lambda x: x['sales'].sum() / x['transactions'].sum()).sort_values(ascending=False)

print("Average Order Size by Store:")
print(avg_order_size_store.head())  

print("\nAverage Order Size by Region:")
print(avg_order_size_region.head()) 

print("\nAverage Order Size by Category:")
print(avg_order_size_category.head())  

Average Order Size by Store:
store_nbr
51   0
42   0
21   0
29   0
52   0
dtype: float64

Average Order Size by Region:
state
Azuay      0
Manabi     0
El Oro     0
Pastaza    0
Los Rios   0
dtype: float64

Average Order Size by Category:
family
GROCERY I   2
BEVERAGES   2
PRODUCE     1
CLEANING    1
DAIRY       0
dtype: float64


In [81]:
df.columns

Index(['date', 'store_nbr', 'family', 'sales', 'onpromotion', 'holiday_type',
       'locale', 'transferred', 'dcoilwtico', 'city', 'state', 'store_type',
       'cluster', 'transactions', 'year', 'month', 'week', 'quarter',
       'day_of_week', 'is_crisis', 'sales_lag_7', 'rolling_mean_7',
       'is_weekend', 'is_holiday', 'promo_last_7_days'],
      dtype='object')

# 📊5. Are there noticeable weekly, monthly, or quarterly seasonality patterns in sales?




In [82]:
day_order = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday']

### What are the trends in sales per day of the week?


In [83]:
avg_sales_by_day = df.groupby('day_of_week')['sales'].mean().reindex(day_order)
avg_sales_by_day

day_of_week
Monday      348
Tuesday     320
Wednesday   331
Thursday    287
Friday      327
Saturday    435
Sunday      465
Name: sales, dtype: float64

### What are the trends in sales per week?

In [84]:
weekly_avg_sales = df.groupby('week')['sales'].mean().reset_index()
weekly_avg_sales

week  sales
0      1    409
1      2    348
2      3    338
3      4    329
4      5    344
5      6    320
6      7    310
7      8    312
8      9    358
9     10    359
10    11    343
11    12    338
12    13    361
13    14    350
14    15    323
15    16    363
16    17    333
17    18    374
18    19    325
19    20    334
20    21    321
21    22    368
22    23    369
23    24    342
24    25    365
25    26    337
26    27    398
27    28    362
28    29    361
29    30    360
30    31    384
31    32    335
32    33    327
33    34    307
34    35    334
35    36    400
36    37    353
37    38    351
38    39    354
39    40    400
40    41    362
41    42    355
42    43    340
43    44    359
44    45    407
45    46    362
46    47    359
47    48    379
48    49    417
49    50    413
50    51    484
51    52    483
52    53    478

### What are the trends in sales per month?

In [85]:
monthly_avg = df.groupby('month')['sales'].mean()
monthly_avg

month
1    342
2    321
3    352
4    341
5    346
6    353
7    376
8    337
9    362
10   362
11   377
12   457
Name: sales, dtype: float64

### What are the trends in sales per quarter?

In [86]:
quarterly_avg = df.groupby('quarter')['sales'].mean()
quarterly_avg

quarter
1   339
2   347
3   359
4   399
Name: sales, dtype: float64

### What are the trends in sales per holiday?

In [87]:
avg_sales_by_holiday = df.groupby('is_holiday')['sales'].mean().reset_index()
avg_sales_by_holiday

is_holiday  sales
0           0    352
1           1    394

### How do weekend sales compare to weekday sales?


In [88]:
df['weekend'] = df['day_of_week'].apply(lambda x: 'Weekend' if x in ['Saturday', 'Sunday'] else 'Weekday')
avg_sales_by_weekend = df.groupby('weekend')['sales'].mean().reset_index()
avg_sales_by_weekend

weekend  sales
0  Weekday    322
1  Weekend    450

# 💸 6. Promotion Impact: What impact do promotions have on sales volume?


### How to compare between promoted and not promoted?

In [99]:
df['promotion_status'] = df['onpromotion'].apply(lambda x: 'On Promotion' if x > 0 else 'Not On Promotion')
print(df['promotion_status'].value_counts())
avg_sales_by_promotion = df.groupby('promotion_status')['sales'].mean().reset_index()
avg_sales_by_promotion

promotion_status
Not On Promotion    2428528
On Promotion         625820
Name: count, dtype: int64


promotion_status  sales
0  Not On Promotion    158
1      On Promotion  1,140

### Is there a cumulative effect of promotions (e.g., last 7 days of promo)?



In [100]:
avg_sales_by_promo_7_days = df.groupby('promo_last_7_days')['sales'].mean().reset_index()
avg_sales_by_promo_7_days

promo_last_7_days  sales
0                    0    198
1                    1    237
2                    2    294
3                    3    338
4                    4    355
..                 ...    ...
905              1,497      5
906              1,521     29
907              1,524  6,825
908              1,545  2,275
909              1,606  6,712

[910 rows x 2 columns]

### Are there specific families or stores where promotions are more effective?

In [101]:
avg_sales_by_family = df.groupby(['family', 'promotion_status'])['sales'].mean().reset_index()
avg_sales_by_family

family  promotion_status  sales
0                   AUTOMOTIVE  Not On Promotion      6
1                   AUTOMOTIVE      On Promotion     13
2                    BABY CARE  Not On Promotion      0
3                    BABY CARE      On Promotion      2
4                       BEAUTY  Not On Promotion      3
..                         ...               ...    ...
60                     PRODUCE      On Promotion  2,435
61  SCHOOL AND OFFICE SUPPLIES  Not On Promotion      1
62  SCHOOL AND OFFICE SUPPLIES      On Promotion     44
63                     SEAFOOD  Not On Promotion     19
64                     SEAFOOD      On Promotion     35

[65 rows x 3 columns]

In [102]:
avg_sales_by_store = df.groupby(['store_nbr', 'promotion_status'])['sales'].mean().reset_index()
avg_sales_by_store

store_nbr  promotion_status  sales
0            1  Not On Promotion    127
1            1      On Promotion    720
2            2  Not On Promotion    178
3            2      On Promotion  1,083
4            3  Not On Promotion    421
..         ...               ...    ...
103         52      On Promotion  1,268
104         53  Not On Promotion     43
105         53      On Promotion    726
106         54  Not On Promotion     80
107         54      On Promotion    677

[108 rows x 3 columns]

# 🌍7. Crisis Impact Analysis

### Crisis Impact by transactions

In [103]:
avg_sales_transactions_crisis = df.groupby('is_crisis')[['sales', 'transactions']].mean().reset_index()
avg_sales_transactions_crisis

is_crisis  sales  transactions
0          0    357         1,557
1          1    495         1,649

### Crisis Impact by Store Type

In [104]:
avg_sales_by_store_crisis = df.groupby(['store_type', 'is_crisis'])['sales'].mean().reset_index()
avg_transactions_by_store_crisis = df.groupby(['store_type', 'is_crisis'])['transactions'].mean().reset_index()


print(avg_sales_by_store_crisis)
print(avg_transactions_by_store_crisis)

  store_type  is_crisis  sales
0          A          0    705
1          A          1    907
2          B          0    325
3          B          1    505
4          C          0    196
5          C          1    268
6          D          0    350
7          D          1    490
8          E          0    268
9          E          1    420
  store_type  is_crisis  transactions
0          A          0         2,859
1          A          1         2,837
2          B          0         1,512
3          B          1         1,702
4          C          0           981
5          C          1         1,062
6          D          0         1,526
7          D          1         1,617
8          E          0         1,017
9          E          1         1,221


### Crisis Impact by promotions

In [105]:
avg_sales_by_promotion_crisis = df.groupby(['is_crisis', 'onpromotion'])['sales'].mean().reset_index()


print(avg_sales_by_promotion_crisis)

     is_crisis  onpromotion  sales
0            0            0    159
1            0            1    470
2            0            2    668
3            0            3    881
4            0            4    990
..         ...          ...    ...
590          1          702  6,825
591          1          710  5,948
592          1          717  6,262
593          1          718  6,712
594          1          720  6,154

[595 rows x 3 columns]


### Crisis Impact by holiday

In [106]:
avg_sales_by_holiday_crisis = df.groupby(['is_crisis', 'is_holiday'])['sales'].mean().reset_index()


print(avg_sales_by_holiday_crisis)

   is_crisis  is_holiday  sales
0          0           0    352
1          0           1    381
2          1           1    495


### Crisis Impact weekly and monthly

In [107]:
avg_sales_by_month_crisis = df.groupby(['is_crisis', 'month'])['sales'].mean().reset_index()
avg_sales_by_week_crisis = df.groupby(['is_crisis', 'week'])['sales'].mean().reset_index()


print(avg_sales_by_month_crisis)
print(avg_sales_by_week_crisis)

    is_crisis  month  sales
0           0      1    342
1           0      2    321
2           0      3    352
3           0      4    321
4           0      5    332
5           0      6    353
6           0      7    376
7           0      8    337
8           0      9    362
9           0     10    362
10          0     11    377
11          0     12    457
12          1      4    523
13          1      5    468
    is_crisis  week  sales
0           0     1    409
1           0     2    348
2           0     3    338
3           0     4    329
4           0     5    344
5           0     6    320
6           0     7    310
7           0     8    312
8           0     9    358
9           0    10    359
10          0    11    343
11          0    12    338
12          0    13    361
13          0    14    350
14          0    15    307
15          0    16    306
16          0    17    305
17          0    18    352
18          0    19    298
19          0    20    332
20          0

### Crisis Impact by transactions, sales

In [ ]:
avg_transactions_sales_crisis = df.groupby('is_crisis')[['transactions', 'sales']].mean().reset_index()


print(avg_transactions_sales_crisis)

   is_crisis  transactions  sales
0          0         1,557    357
1          1         1,649    495


### Crisis Impact by family

In [111]:
avg_sales_by_family_crisis = df.groupby(['family', 'is_crisis'])['sales'].mean().reset_index()


print(avg_sales_by_family_crisis)

                        family  is_crisis  sales
0                   AUTOMOTIVE          0      6
1                   AUTOMOTIVE          1      7
2                    BABY CARE          0      0
3                    BABY CARE          1      0
4                       BEAUTY          0      4
..                         ...        ...    ...
61                     PRODUCE          1  2,265
62  SCHOOL AND OFFICE SUPPLIES          0      3
63  SCHOOL AND OFFICE SUPPLIES          1      9
64                     SEAFOOD          0     22
65                     SEAFOOD          1     24

[66 rows x 3 columns]


### Crisis Impact by city and state

In [112]:
avg_sales_by_city_crisis = df.groupby(['city', 'is_crisis'])['sales'].mean().reset_index()


avg_sales_by_state_crisis = df.groupby(['state', 'is_crisis'])['sales'].mean().reset_index()


print(avg_sales_by_city_crisis)
print(avg_sales_by_state_crisis)

             city  is_crisis  sales
0          Ambato          0    363
1          Ambato          1    429
2        Babahoyo          0    319
3        Babahoyo          1    417
4         Cayambe          0    509
5         Cayambe          1    636
6          Cuenca          0    293
7          Cuenca          1    434
8           Daule          0    344
9           Daule          1    505
10      El Carmen          0    199
11      El Carmen          1    269
12     Esmeraldas          0    295
13     Esmeraldas          1    348
14       Guaranda          0    234
15       Guaranda          1    305
16      Guayaquil          0    275
17      Guayaquil          1    400
18         Ibarra          0    205
19         Ibarra          1    267
20      Latacunga          0    190
21      Latacunga          1    248
22       Libertad          0    275
23       Libertad          1    389
24           Loja          0    340
25           Loja          1    378
26        Machala          0

### Crisis Impact on Rolling Mean and Lagged Sales

In [113]:
avg_sales_lag_7_crisis = df.groupby('is_crisis')['sales_lag_7'].mean().reset_index()
avg_rolling_mean_7_crisis = df.groupby('is_crisis')['rolling_mean_7'].mean().reset_index()


print(avg_sales_lag_7_crisis)
print(avg_rolling_mean_7_crisis)

   is_crisis  sales_lag_7
0          0          355
1          1          492
   is_crisis  rolling_mean_7
0          0             356
1          1             495


### Crisis and Store Cluster Performance

In [114]:
avg_sales_by_cluster_crisis = df.groupby(['cluster', 'is_crisis'])['sales'].mean().reset_index()
avg_transactions_by_cluster_crisis = df.groupby(['cluster', 'is_crisis'])['transactions'].mean().reset_index()


print(avg_sales_by_cluster_crisis)
print(avg_transactions_by_cluster_crisis)

    cluster  is_crisis  sales
0         1          0    325
1         1          1    432
2         2          0    259
3         2          1    390
4         3          0    194
5         3          1    254
6         4          0    297
7         4          1    338
8         5          0  1,113
9         5          1  1,510
10        6          0    341
11        6          1    533
12        7          0    138
13        7          1    221
14        8          0    644
15        8          1    896
16        9          0    274
17        9          1    351
18       10          0    255
19       10          1    375
20       11          0    602
21       11          1    814
22       12          0    322
23       12          1    512
24       13          0    322
25       13          1    548
26       14          0    708
27       14          1    853
28       15          0    198
29       15          1    257
30       16          0    236
31       16          1    424
32       1